src/persist-client/src/write.rs

```diff
@@ -329,56 +437,224 @@ where
         }
     }
 
-    fn encode_batch<SB, KB, VB, TB, DB, I>(
-        desc: &Description<T>,
-        updates: I,
-    ) -> Result<Option<Vec<u8>>, InvalidUsage<T>>
+    /// Returns a [BatchBuilder] that can be used to write a batch of updates
+    /// to blob storage which can then be appended to this shard using
+    /// [Self::compare_and_append_core] or [Self::append_core].
+    ///
+    /// It is correct to create an empty batch, which allows for downgrading
+    /// `upper` to communicate progress. (see [Self::compare_and_append_core] or
+    /// [Self::append_core])
+    ///
+    /// The builder uses a bounded amount of memory, even when the number of
+    /// updates is very large. Individual records, however, should be small
+    /// enough that we can reasonably chunk them up: O(KB) is definitely fine,
+    /// O(MB) come talk to us.
+    // NOTE: This function is async right now, even though it doesn't need it.
```

Comment:
nit: empty line between these

I'm down with reserving async for the future, but curious what you're imagining here? in my mind, `BatchBuilder::add` being async should be sufficient, but I very well could be missing something